torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -518,6 +518,19 @@ class ShapePropagator {
         }
         return;
       }
+      case prim::unchecked_unwrap_optional: {
```

Comment:
Input to unchecked_unwrap_optional should _always_ be an optional type. Things will break if it isn't

Reply:
Well, no. If we specialize optionals, it might be element type instead.

Reply:
I changed it to test whether the input is a subtype of the output and if so, take the input type. This allows `Optional[Tensor]`s that have been specialized to `DimensionedTensorType` to have the more precise output type.

Reply:
I think this is ok now.